### PR TITLE
AMDGPU: Generalize instruction shrinking code

### PR DESCRIPTION
**Pull request by arsenm**
Try to avoid referring to specific operand names, except in the special case. The special case for hasNamedOperand(Op32, sdst) seems to have been dead code.

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 1]
```diff
@@ -4486,46 +4486,47 @@ static void copyFlagsToImplicitVCC(MachineInstr &MI,
...
-      fixImplicitOperands(*Inst32);
-      copyFlagsToImplicitVCC(*Inst32, *Src2);
+  int Idx = MI.getNumExplicitDefs();
+  for (const MachineOperand &Use : MI.explicit_uses()) {
+    int OpTy = MI.getDesc().operands()[Idx++].OperandType;
+    if (OpTy == AMDGPU::OPERAND_INPUT_MODS || OpTy == MCOI::OPERAND_IMMEDIATE)
+      continue;
+
+    if (&Use == Src2) {
+      int Op32Src2Idx = AMDGPU::getNamedOperandIdx(Op32, AMDGPU::OpName::src2);
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4512]: Nit: inline `Op32Src2Idx` into its only use?

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp [hunk 2]
```diff
@@ -4486,46 +4486,47 @@ static void copyFlagsToImplicitVCC(MachineInstr &MI,
...
-      // replaced with an implicit read of vcc or vcc_lo. The implicit read
-      // of vcc was already added during the initial BuildMI, but we
-      // 1) may need to change vcc to vcc_lo to preserve the original register
-      // 2) have to preserve the original flags.
-      fixImplicitOperands(*Inst32);
-      copyFlagsToImplicitVCC(*Inst32, *Src2);
+  int Idx = MI.getNumExplicitDefs();
+  for (const MachineOperand &Use : MI.explicit_uses()) {
+    int OpTy = MI.getDesc().operands()[Idx++].OperandType;
+    if (OpTy == AMDGPU::OPERAND_INPUT_MODS || OpTy == MCOI::OPERAND_IMMEDIATE)
```
Comment (jayfoad) [inline llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4508]: What cases does `OPERAND_IMMEDIATE` cover? Does this mean we don't copy any immediate operands, and if so why not?

**Review by jayfoad — COMMENTED** [review]
Seems fine if it works, just a couple of nits inline.

File: llvm/lib/Target/AMDGPU/SIInstrInfo.cpp (hunk 2 above)
Comment (arsenm) [reply llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:4508]: It's the default case for custom operands. The most common cases are clamp and omod

**Review by jayfoad — APPROVED** [review]

**Pull request merged**